IPIP/0298-denylists-for-ipfs-nodes-and-gateways.md

```diff
@@ -0,0 +1,243 @@
+      Type: 3, // Type.SHA256_CID_V1
+      Content: "9056e0f9948c942c16af3564af56d4bb96b6203ad9ccd3425ec628bcd843cc39",
+      Description: "sensitive cid that needs to be blocked"
+    }
+  ]
+}
+```
+
+### Action
+
+Action is the suggested action to be taken by the IPFS node for content present in the list. Actions are defined as follow
+
+| Value | Code | Description |
+|:------|:-----|:------------|
+|0|UNDEFINED|No action to be taken|
+|1|ALLOW|Content in the list should be allowed|
+|2|BLOCK|Content in the list should be blocked|
+|3-255|Unassigned|Reserved for future use|
+|256-512|INTERNAL|This range can be used for private actions|
+
+### Status Code
+
+Status Code is a default HTTP status code. This is meant to be served by [IPFS HTTP Gateways](../http-gateways/README.md) upon a matching item.
+
+This is the default HTTP Status Code for the list. If an Entry has a Status Code specified, the entry code takes priority.
+
+For example, a status code can be [410 Gone](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#410-gone); [451 Unavailable For Legal Reasons](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#451-unavailable-for-legal-reasons) or `200 OK` for allowed entry.
+
+### Entries
+
+Entries is a list of content for which Action has to be taken for upon matching. Matching happens based on the type.
+
+Matching logic may differ depending on the implementation. It is recommended to pass the request path, as well as the CIDs of the composing DAG to the rule engine.
+
+#### Type and Content
+
+| Value | Code | Description | Content | Match on |
+|:------|:-----|:------------|:--------|:---------|
+|0|CID_V1|CIDv1 of the content at the root or its children|[Multibase](https://github.com/multiformats/multibase/blob/master/multibase.csv) encoded CIDv1 as a string|Requested CID v1 at the root or its children|
+|1|ABSOLUTE_PATH|Absolute path of a content from the root of a IPLD node|Path as a string|Requested path (either via https:// or ipfs://) starts with|
+|2|PARTIAL_PATH|Part of a path of a content|Path as a string|Requested path contains|
+|3|SHA256_CID_V1|SHA256 hased CID v1. This matches the [Double Hashing IPIP-334](https://github.com/ipfs/specs/pull/334).|SHA256 of the CIDv1 as a string|Requested CID v1 SHA256|
+|4-255|Unassigned||||
+|256-512|INTERNAL|This range can be used for private content type|||
+
+> **Note on `SHA256_CID_V1` type**
+>
+> The main difference between non-hashed entries and hashed ones is that the CIDs or content paths in the entry will be hashed and no plaintext is shown in the list. Following the [bad bits](https://badbits.dwebops.pub/), each CID or content path is `sha256()` hashed, so it's easy to determine one way but not the other. The hashed entries are designed to store sensitive blocking items and prevent creating an easily accessible list of sensitive content.
+>
+> Before the hashing, all CIDv0 in `CID_V1` field is converted to CIDv1 for consistency.
```

Comment:
I'm super late to the party, and this is maybe fixed elsewhere, but I wanted to +1 @dokterbob's suggestion about salting the double hashes (and making sure that the salt doesn't match the one in the double-hashed DHT). Otherwise you're handing a treasure map to bad actors.

There's perhaps a broader discussion about the privacy properties that we need from such a system. These are all some form of private set intersection (PSI), where double hashing is the "naive" case — there are much more sophisticated techniques. (Fancier _does not_ nessesarily mean better, but may be worth exploring)